### PR TITLE
Build, deploy and test the Quote Lifecycle Simulator

### PR DESCRIPTION
**Pull request by Kinshuk1993**
Create a pipeline to build/deploy the quote lifecycle simulator
- Refactor the common tasks yaml for build and to deploy - move to a common directory
- Get an inline task using `taskSpec` in the pipeline for eei for build and deploy
- Create a pipeline yaml (include a task to clone the git repository as the first task call)
- Create a pipelinerun yaml
- Create required RBACs for pipeline tasks
- Apply the created pipeline resources from the prereqs of EEI
- Build Simulator task, use Buildah to build an image from the Dockerfile
- Deploy Simulator task, run the script to deploy the app with 0 replicas
- Write instructions for how to run the pipeline to build/deploy the simulator and how to scale it up/down.
- Test DDD after the refactoring
- Test EEI with Postgres debezium connector
- Test the updated 1-click
- Test the pipelineruns for EEI and DDD after 1-click installs are done

### CODE REVIEW

File: EventEnabledInsurance/prereqs.sh [hunk 1]
```diff
@@ -158,14 +222,168 @@ stringData:
...
+
+  echo -e "\nINFO: Waiting up to 5 minutes for the pipelinerun for the quote lifecycle simulator app to finish. Waited ${time} minute(s)."
+  time=$((time + 1))
+  sleep 60
+done
+
+echo -e "\n$tick INFO: The pipelinerun for the quote lifecycle simulator app has completed successfully, going ahead to delete the pipelinerun for it to delete the pods and the pvc\n"
+oc get pipelinerun -n $namespace $(oc get pipelinerun -n $namespace | grep $PIPELINERUN_UID | awk '{print $1}')
+
+echo -e "\n$tick INFO: The quote lifecycle simulator application has been deployed, but with zero replicas. To scale up the pods, run the command 'kubectl scale --replicas=<expected_replica_num> deployment simulator-eei'\n"
```
Comment (DanRoseus) [inline EventEnabledInsurance/prereqs.sh:386]:
```suggestion
echo -e "\n$tick INFO: The quote lifecycle simulator application has been deployed, but with zero replicas. To start the simulator run the command 'oc scale deployment/quote-simulator-eei --replicas=1'\n"
```

File: EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh [hunk 2]
```diff
@@ -20,10 +20,10 @@
 #     ./prereqs.sh
 #
 #   With overridden values
-#     ./prereqs.sh -n <namespace> -t <TICK_MILLIS> -m <MOBILE_TEST_ROWS>
+#     ./prereqs.sh -n <namespace>
```
Comment (DanRoseus) [inline EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh:23]:
```suggestion
#     ./deploy-simulator.sh -n <namespace>
```

File: EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh [hunk 3]
```diff
@@ -38,20 +38,14 @@ PG_PORT=5432
...
-  t)
-    TICK_MILLIS="$OPTARG"
-    ;;
-  m)
-    MOBILE_TEST_ROWS="$OPTARG"
-    ;;
   \?)
     usage
-    exit
+    # exit
```
Comment (DanRoseus) [inline EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh:48]: Any reason this is commented out?

File: EventEnabledInsurance/QuoteLifecycleSimulator/readme.md [hunk 4]
```diff
@@ -70,3 +70,27 @@ Run the application:
...
+export namespace=<yourNamespace>
+```
+- Now run the prereqs script from the `EventEnabledInsurance` directory with the following command:
+```
+./prereqs.sh -n $namespace -b $BRANCH -r $FORKED_REPO
+```
+- The prereqs will set up the dependencies, secrets, the database and configures the pipeline to deploy the simulator app with 0 replicas.
+- To scale up/down the replicas for the deployment:
+```
+kubectl scale --replicas=<expected_replica_num> deployment simulator-eei
```
Comment (DanRoseus) [inline EventEnabledInsurance/QuoteLifecycleSimulator/readme.md:95]:
```suggestion
oc scale deployment/quote-simulator-eei --replicas=<0/1>
```

File: EventEnabledInsurance/prereqs.sh [hunk 5]
```diff
@@ -158,14 +222,184 @@ stringData:
...
+else
+  echo -e "\n$cross ERROR: Failed to create rolebinding for creating deployment in the '$namespace' namespace for simulator app"
+  exit 1
+fi #simulator-rolebindings.yaml
+
+echo -e "\n----------------------------------------------------------------------------------------------------------------------------------------------------------\n"
+
+time=0
+echo "INFO: Waiting for upto 10 minutes for git-clone cluster task to be available before creating the pipeline and the pipelinerun..."
+while [ "$($TKN clustertask ls | grep git-clone | awk '{print $1}')" != "git-clone" ]; do
```
Comment (DanRoseus) [inline EventEnabledInsurance/prereqs.sh:306]: Why not just check if `oc get clustertask git-clone` succeeds? Saves messing around with grep and awk, and is safer. If there's a cluster task with a string that contains `git-clone` (I.e. `my-git-clone-task`), then the grep/awk won't work.

**Review by DanRoseus — CHANGES_REQUESTED** [review]

File: EventEnabledInsurance/prereqs.sh (hunk 1 above)
Comment (Kinshuk1993) [reply EventEnabledInsurance/prereqs.sh]: Done

File: EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh (hunk 3 above)
Comment (Kinshuk1993) [reply EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh:48]: The usage already has an exit inside it, so removed this exit.

File: EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh (hunk 2 above)
Comment (Kinshuk1993) [reply EventEnabledInsurance/QuoteLifecycleSimulator/deploy-simulator.sh:23]: Done

File: EventEnabledInsurance/QuoteLifecycleSimulator/readme.md (hunk 4 above)
Comment (Kinshuk1993) [reply EventEnabledInsurance/QuoteLifecycleSimulator/readme.md]: Done

File: EventEnabledInsurance/prereqs.sh (hunk 5 above)
Comment (Kinshuk1993) [reply EventEnabledInsurance/prereqs.sh:306]: I was wondering that would the cluster task be named anything else at all apart from git-clone considering that it is created by the ocp pipelines?

**Comment by Kinshuk1993** [comment]
- [wait-clustertask-gitclone-log.txt](https://github.com/IBM/cp4i-deployment-samples/files/5188802/wait-clustertask-gitclone-log.txt)

File: EventEnabledInsurance/prereqs.sh (hunk 5 above)
Comment (Kinshuk1993) [reply EventEnabledInsurance/prereqs.sh]: Done

```
INFO: Waiting for upto 10 minutes for git-clone cluster task to be available before creating the pipeline and the pipelinerun...
Error from server (NotFound): clustertasks.tekton.dev "git-clones" not found
git-clone                                22 minutes ago

INFO: The cluster task ''git-clone' is not yet available, waiting for upto 10 minutes. Waited 0 minute(s).

INFO: Cluster task 'git-clone' is now available

git-clone                                23 minutes ago

----------------------------------------------------------------------------------------------------------------------------------------------------------

INFO: Creating the pipeline to build and deploy the simulator app in 'cp4i1' namespace
pipeline.tekton.dev/eei-simulator-pipeline configured

✅ INFO: Successfully applied the pipeline to build and deploy the simulator app in in 'cp4i1' namespace

----------------------------------------------------------------------------------------------------------------------------------------------------------
```

**Review by DanRoseus — APPROVED** [review]
Looks good!

**Pull request merged**